### ISSUE TITLE
Wrong type inference with boolean assignment

### ISSUE THREAD

**Issue by vhiairrassary** [microsoft/TypeScript#48268]
# Bug Report

### 🔎 Search Terms

- Typescript logical or assignment inference issue

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.2#code/DYUwLgBAHhC8EDMCGwDOIDcAoA2gRgBoIAmAXQDoEB7AJwFEkBjACwApWBKOAPggG8sEIdAgAfUfDA0ArpiwBfDliyMqAO1RVQ5YFQDmrKB2yqNWkDv2G4sSTJDGsQA)

### 💻 Code

`x` is inferred to be `false` while it should be `boolean`;
```typescript
let x = false;
[1, 2].forEach(() => {
    x ||= true;
})

console.log(x);
console.log(x === true); // Error: This condition will always return 'false' since the types 'false' and 'true' have no overlap.
```

The following snippet does not raise an error
```typescript
let x = false;
x ||= true;

console.log(x);
console.log(x === true);
```

**Comment by Josh-Cena** [comment]
Sounds like #9998 🤔 It's because we can't be sure that the array callbacks are immediately executed before logging `x === true`, so the type of `x` is still `false`. You need `let x = false as boolean;` or better, `let x = false satisfies boolean;` in the next version

**Comment by cypherfunc** [comment]
> You need `let x = false as boolean;` or better, `let x = false satisfies boolean;` in the next version

or better yet, `let x: boolean = false;`. Why cast when you can use an explicit type? (super excited for `satisfies` though)

**Comment by Josh-Cena** [comment]
Well... If I could I would:

https://www.typescriptlang.org/play?#code/DYUwLgBAhgXBBGB7RooDsIF4IDMrAGcQBuAKAEscIAKKLTbMAJwFcQBKCAb1IggGNEaAihAA6YIgDm1AEQB1ABYBPCGkRgA-LPakAvqSA

**Comment by cypherfunc** [comment]
> Well... If I could I would:
> 
> https://www.typescriptlang.org/play?#code/DYUwLgBAhgXBBGB7RooDsIF4IDMrAGcQBuAKAEscIAKKLTbMAJwFcQBKCAb1IggGNEaAihAA6YIgDm1AEQB1ABYBPCGkRgA-LPakAvqSA

.....what the heck?😕
I assume there's a specific reason for TS to act that way, but it makes no sense to me.

**Comment by Josh-Cena** [comment]
My guess is that it's still #9998, that type narrowing is done too optimistically. `let x: boolean = false;` is even presented as the first use-case of `satisfies`😅

**Comment by RyanCavanaugh** [comment]
Initialization is narrowing, otherwise if you wrote something like
```ts
type SomeSuperCoolType = { a: string, b: number } | undefined;

let v: SomeSuperCoolType = {
  a: "hello",
  b: 32
};
console.log(a.v);
```
we'd issue "`a` is possibly undefined", which is daft
